runway/hooks/staticsite/auth_at_edge/callback_url_retriever.py

```diff
@@ -0,0 +1,63 @@
+"""Callback URL Retriever.
+
+Dependency pre-hook responsible for ensuring correct
+callback urls are retrieved or a temporary one is used in it's place.
+"""
+
+import logging
+
+from typing import Any, Dict, Optional  # pylint: disable=unused-import
+
+from runway.cfngin.providers.base import BaseProvider  # pylint: disable=unused-import
+from runway.cfngin.context import Context  # noqa pylint: disable=unused-import
+from runway.cfngin.session_cache import get_session
+
+LOGGER = logging.getLogger(__name__)
+
+
+def get(context,  # pylint: disable=unused-argument
+        provider,
+        **kwargs
+       ):  # noqa: E124
+    # type: (Context, BaseProvider, Optional[Dict[str, Any]]) -> Dict
+    """Retrieve the callback URLs for User Pool Client Creation.
+
+    When the User Pool is created a Callback URL is required. During a post
+    hook entitled ``client_updater`` these Callback URLs are updated to that
+    of the Distribution. Before then we need to ensure that if a Client
+    already exists that the URLs for that client are used to prevent any
+    interuption of service during deploy.
+
+    Args:
+        context (:class:`runway.cfngin.context.Context`): The context
+            instance.
+        provider (:class:`runway.cfngin.providers.base.BaseProvider`):
+            The provider instance
+
+    Keyword Args:
+        user_pool_id (str): The ID of the User Pool to check for a client
+    """
+    session = get_session(provider.region)
+    cognito_client = session.client('cognito-idp')
+    context_dict = {}
+    context_dict['callback_urls'] = ['https://example.tmp']
+
+    try:
+        clients = cognito_client.list_user_pool_clients(
+            UserPoolId=kwargs['user_pool_id']
+        )['UserPoolClients']
+
+        client = next((c for c in clients if c['ClientName'].startswith('AuthAtEdge')), None)
```

Comment:
Fair, I'll see what I can do with the name that's generated

Reply:
I'm not really sure how we'll do this effectively, but am open to suggestions. 

At the point in time where this is called, we have no data for the application other than a UserPool ARN/ID. We could use those to hash in the client name, but we'll end up with duplicates if we're using the same User Pool and end up being in the same spot we are now. I'm not really sure how we can differentiate the correct ClientID in this case since we're in the process of trying to poll for that info.

Is there a way with CFNgin for me to say "If we're on an update step, and a User Pool Client already exists that was created by CFNgin, then give me that ClientID?".

If there is a way to do the above then I should probably refactor this whole check to get that info and just poll the names from that specific one, otherwise, return the `example.tmp` domain.

Reply:
> Is there a way with CFNgin for me to say "If we're on an update step, and a User Pool Client already exists that was created by CFNgin, then give me that ClientID?".

Without SSM*

Reply:
After discussing with @troyready decided we would poll the stack information instead to retrieve the info (when available)